File: python/seal_gate/detectors/evidence_gap_detector.py
import re
from ..types import SealIssue, SealEvidence, make_issue, RISK_ORDER


def detect_evidence_gaps(claims: list, evidence: SealEvidence, risk_level: str = 'MEDIUM') -> dict:
    issues: list[SealIssue] = []
    missing: list[str] = []
    deductions = 0

    for claim in claims:
        if claim.type == 'test_result_claim':
            has_log = bool(evidence.test_log.strip())
            has_ref = any(r.get('type') == 'command' and re.search(r'test|spec|jest|vitest|pytest', r.get('command', ''), re.I) for r in evidence.references)
            if not has_log and not has_ref:
                issues.append(make_issue(type='MISSING_EVIDENCE', severity='HIGH', layer='L3', source='core', rule_id='E001', trust_deduction=20, evidence=f'No test log for claim: "{claim.text}"', required_fix='Provide test_log or command evidence envelope with test output'))
                missing.append(f'test_log required for: {claim.text}')
                deductions += 20

        elif claim.type == 'build_claim':
            has_log = bool(evidence.build_log.strip())
            has_ref = any(r.get('type') == 'command' and re.search(r'build|compile|make|gradle|maven|tsc', r.get('command', ''), re.I) for r in evidence.references)
            if not has_log and not has_ref:
                issues.append(make_issue(type='MISSING_EVIDENCE', severity='MEDIUM', layer='L3', source='core', rule_id='E002', trust_deduction=15, evidence=f'No build log for claim: "{claim.text}"', suggested_fix='Provide build_log or command evidence'))
                deductions += 15

        elif claim.type == 'risk_claim':
            has_sec = bool(evidence.test_log.strip()) and bool(re.search(r'security|auth|unauthorized|pentest|vuln', evidence.test_log, re.I))
            has_ref = any(r.get('type') == 'text' and re.search(r'security', r.get('label', ''), re.I) for r in evidence.references)
            if not has_sec and not has_ref:
                issues.append(make_issue(type='MISSING_EVIDENCE', severity='HIGH', layer='L3', source='core', rule_id='E003', trust_deduction=20, evidence=f'No security reasoning for claim: "{claim.text}"', required_fix='Provide security test log or text envelope'))
                deductions += 20

        elif claim.type == 'production_claim' and RISK_ORDER.index(risk_level) >= RISK_ORDER.index('MEDIUM'):
            has_any = bool(evidence.references) or bool(evidence.test_log.strip())
            if not has_any:
                issues.append(make_issue(type='MISSING_EVIDENCE', severity='CRITICAL', layer='L3', source='core', rule_id='E004', trust_deduction=20, evidence=f'No evidence for production claim: "{claim.text}"', required_fix='Provide evidence envelopes'))
                missing.append(f'production evidence required for: {claim.text}')
                deductions += 20

    return {'issues': issues, 'missing_evidence': missing, 'trust_deductions': deductions}

File: src/detectors/evidence-gap-detector.ts
import { Claim, SealEvidence, SealIssue, EvidenceEnvelope } from '../types.ts'
import { makeIssue } from '../types.ts'

function hasMemoryEnvelope(references: EvidenceEnvelope[]): boolean {
  return references.some(r => r.type === 'memory' || (r.type === 'text' && ('label' in r) && /alaya|retrieved/i.test(r.label)))
}

export interface EvidenceGapResult {
  issues: SealIssue[]
  missing_evidence: string[]
  trust_deductions: number
}

const RISK_ORDER = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']

export class EvidenceGapDetector {
  static detect(claims: Claim[], evidence: SealEvidence, risk_level = 'MEDIUM'): EvidenceGapResult {
    const issues: SealIssue[] = []
    const missing_evidence: string[] = []
    let trust_deductions = 0

    for (const claim of claims) {
      if (claim.type === 'test_result_claim') {
        const hasLog = evidence.test_log.trim().length > 0
        const hasCommandEvidence = evidence.references.some(r =>
          r.type === 'command' && /test|spec|jest|vitest|pytest/i.test(r.command)
        )
        if (!hasLog && !hasCommandEvidence) {
          issues.push(makeIssue({
            type: 'MISSING_EVIDENCE', severity: 'HIGH', layer: 'L3', source: 'core',
            rule_id: 'E001', trust_deduction: 20,
            evidence: `No test log for claim: "${claim.text}"`,
            required_fix: 'Provide test_log or command evidence envelope with test output',
          }))
          missing_evidence.push(`test_log required for: ${claim.text}`)
          trust_deductions += 20
        }
      }

      if (claim.type === 'build_claim') {
        const hasLog = evidence.build_log.trim().length > 0
        const hasCommandEvidence = evidence.references.some(r =>
          r.type === 'command' && /build|compile|make|gradle|maven|tsc/i.test(r.command)
        )
        if (!hasLog && !hasCommandEvidence) {
          issues.push(makeIssue({
            type: 'MISSING_EVIDENCE', severity: 'MEDIUM', layer: 'L3', source: 'core',
            rule_id: 'E002', trust_deduction: 15,
            evidence: `No build log for claim: "${claim.text}"`,
            suggested_fix: 'Provide build_log or command evidence envelope with build output',
          }))
          trust_deductions += 15
        }
      }

      if (claim.type === 'risk_claim') {
        const hasSecurityEvidence = evidence.test_log.trim().length > 0 &&
          /security|auth|unauthorized|pentest|vuln/i.test(evidence.test_log)
        const hasRefEvidence = evidence.references.some(r =>
          r.type === 'text' && /security/i.test('label' in r ? r.label : '')
        )
        if (!hasSecurityEvidence && !hasRefEvidence) {
          issues.push(makeIssue({
            type: 'MISSING_EVIDENCE', severity: 'HIGH', layer: 'L3', source: 'core',
            rule_id: 'E003', trust_deduction: 20,
            evidence: `No security reasoning for claim: "${claim.text}"`,
            required_fix: 'Provide security test log or security-reasoning text envelope',
          }))
          trust_deductions += 20
        }
      }

      if (claim.type === 'production_claim' && RISK_ORDER.indexOf(risk_level) >= RISK_ORDER.indexOf('MEDIUM')) {
        const hasAnyEvidence = evidence.references.length > 0 || evidence.test_log.trim().length > 0
        if (!hasAnyEvidence) {
          issues.push(makeIssue({
            type: 'MISSING_EVIDENCE', severity: 'CRITICAL', layer: 'L3', source: 'core',
            rule_id: 'E004', trust_deduction: 20,
            evidence: `No evidence for production-readiness claim: "${claim.text}"`,
            required_fix: 'Provide evidence envelopes demonstrating production readiness',
          }))
          missing_evidence.push(`production evidence required for: ${claim.text}`)
          trust_deductions += 20
        }
      }
    }

    return { issues, missing_evidence, trust_deductions }
  }
}
